src/main/java/com/puppycrawl/tools/checkstyle/XMLLogger.java

```diff
@@ -121,40 +148,107 @@ public void auditFinished(AuditEvent event) {
 
     @Override
     public void fileStarted(AuditEvent event) {
-        writer.println("<file name=\"" + encode(event.getFileName()) + "\">");
+        fileMessages.put(event.getFileName(), new FileMessages());
     }
 
     @Override
     public void fileFinished(AuditEvent event) {
+        final String fileName = event.getFileName();
+        final FileMessages errorEvents = fileMessages.get(fileName);
+
+        synchronized (writerLock) {
+            if (errorEvents != null) {
+                writeFileOpeningTag(fileName);
+
+                for (AuditEvent errorEvent : errorEvents.errors) {
+                    writeFileError(errorEvent);
+                }
+                for (Throwable exception : errorEvents.exceptions) {
+                    writeException(exception);
+                }
+            }
+
+            writeFileClosingTag();
```

Comment:
Shouldn't file open tag be on the same level as file close tag?
If `errorEvents` is null, you will write a close tag without any open tag.